zkevm-circuits/src/keccak_circuit/keccak_packed.rs

```diff
@@ -0,0 +1,1770 @@
+        let packed = split(
+            meta,
+            &mut cell_values,
+            &mut cb,
+            squeeze_from.clone(),
+            0,
+            8,
+            false,
+        );
+        transform(
+            "squeeze unpack",
+            meta,
+            &mut cell_values,
+            &mut lookup_counter,
+            packed,
+            pack_unpack_table
+                .into_iter()
+                .rev()
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap(),
+        );
+
+        meta.create_gate("input checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+            cb.require_boolean(
+                "boolean is_final",
+                meta.query_advice(is_final, Rotation::cur()),
+            );
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        meta.create_gate("first row", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+            cb.require_equal(
+                "is_final needs to be enabled on the first row",
+                meta.query_advice(is_final, Rotation::cur()),
+                1.expr(),
+            );
+            cb.gate(meta.query_selector(q_first))
+        });
+
+        meta.create_gate("round", |meta| {
+            cb.gate(meta.query_fixed(q_round, Rotation::cur()))
+        });
+
+        s = pre_s;
+
+        meta.create_gate("absorb", |meta| {
+            let mut cb = BaseConstraintBuilder::new(3);
+
+            let is_not_final = not::expr(meta.query_advice(is_final, Rotation::cur()));
+
+            let absorb_positions = get_absorb_positions();
+            let mut input_slice = 0;
+            for j in 0..5 {
+                for i in 0..5 {
+                    if absorb_positions.contains(&(i, j)) {
+                        cb.condition(is_not_final.clone(), |cb| {
+                            cb.require_equal(
+                                "absorb verify input",
+                                absorb_from_next_expr[input_slice].clone(),
+                                s[i][j].clone(),
+                            );
+                        });
+                        cb.require_equal(
+                            "absorb result copy",
+                            select::expr(
+                                is_not_final.clone(),
+                                absorb_result_next_expr[input_slice].clone(),
+                                absorb_data_next_expr[input_slice].clone(),
+                            ),
+                            s_next[i][j].clone(),
+                        );
+                        input_slice += 1;
+                    } else {
+                        cb.require_equal(
+                            "absorb state copy",
+                            s[i][j].clone() * is_not_final.clone(),
+                            s_next[i][j].clone(),
+                        );
```

Comment:
Maybe so. I created a helper function because this is used in a couple of places and was easier to change it (I actually had this wrong for some time).